pkg/userdata/input.go

```diff
@@ -0,0 +1,70 @@
+package userdata
+
+import (
+	"fmt"
+	"strings"
+
+	bootstrapv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/util/secret"
+)
+
+// ControlPlaneInput defines the context to generate a controlplane instance user data.
+type EtcdPlaneInput struct {
+	BaseUserData
+	secret.Certificates
+	EtcdadmArgs
+
+	EtcdadmInitCommand string
+}
+
+// EtcdPlaneJoinInput defines context to generate etcd instance user data for etcd plane node join.
+type EtcdPlaneJoinInput struct {
+	BaseUserData
+	secret.Certificates
+	EtcdadmArgs
+
+	EtcdadmJoinCommand string
+	JoinAddress        string
+	Version            string
+}
+
+// BaseUserData is shared across all the various types of files written to disk.
+type BaseUserData struct {
+	Header              string
+	PreEtcdadmCommands  []string
+	PostEtcdadmCommands []string
+	AdditionalFiles     []bootstrapv1.File
+	WriteFiles          []bootstrapv1.File
+	Users               []bootstrapv1.User
+	NTP                 *bootstrapv1.NTP
+	DiskSetup           *bootstrapv1.DiskSetup
+	Mounts              []bootstrapv1.MountPoints
+	ControlPlane        bool
+	SentinelFileCommand string
+}
+
+type EtcdadmArgs struct {
+	Version         string
+	ImageRepository string
+}
+
+func (args *EtcdadmArgs) Flags() []string {
+	flags := make([]string, 0, 3)
+	flags = append(flags, "--init-system kubelet")
```

Comment:
This will run etcd on pods when using cloud-init (I haven't tested it)
We can move it a different PR